src/contracts/caching_log_query.rs

```diff
@@ -0,0 +1,214 @@
+    EmptyBlockIndex,
+    #[error("empty transaction index")]
+    EmptyTransactionIndex,
+    #[error("empty log index")]
+    EmptyLogIndex,
+}
+
+impl CachingLogQuery {
+    pub fn new(provider: Arc<ProviderStack>, filter: &Filter) -> Self {
+        Self {
+            provider,
+            filter: filter.clone(),
+            page_size: 10000,
+            cache_blocks_delay: 0,
+            database: None,
+        }
+    }
+
+    /// set page size for pagination
+    pub const fn with_page_size(mut self, page_size: u64) -> Self {
+        self.page_size = page_size;
+        self
+    }
+
+    pub const fn with_blocks_delay(mut self, cache_blocks_delay: u64) -> Self {
+        self.cache_blocks_delay = cache_blocks_delay;
+        self
+    }
+
+    #[allow(clippy::missing_const_for_fn)]
+    pub fn with_database(mut self, database: Option<Arc<Database>>) -> Self {
+        self.database = database;
+        self
+    }
+
+    pub fn into_stream(self) -> impl Stream<Item = Result<Log, Error<ProviderError>>> {
+        try_stream! {
+            let last_block = self.get_block_number().await?;
+
+            let cached_events = self.load_db_logs().await?;
+            for log in cached_events {
+                yield serde_json::from_str(log.get()).map_err(Error::Parse)?;
+            }
+
+            let paginator = Paginator::new(last_block, self.filter.clone(), self.page_size);
+            for page in paginator {
+                let new_events = self.fetch_page(page.filter).await?;
+                for raw_log in new_events {
+                    let log: Log = serde_json::from_str(raw_log.get()).map_err(Error::Parse)?;
+                    if self.should_cache(&log, last_block) {
+                        self.cache_log(raw_log, &log).await?;
+                    }
+                    yield log;
+                }
+            }
+        }
+    }
+
+    async fn get_block_number(&self) -> Result<LastBlock, Error<ProviderError>> {
+        let provider = self.provider.provider();
+        let last_eth_block = provider
+            .get_block_number()
+            .await
+            .map_err(Error::LoadLastBlock)?;
+        let last_db_block: u64 = match &self.database {
+            Some(database) => database.get_block_number().await? as u64,
+            None => 0,
+        };
+
+        Ok(LastBlock {
+            eth: last_eth_block,
+            db:  U64([last_db_block]),
+        })
+    }
+
+    async fn load_db_logs(&self) -> Result<Vec<Box<RawValue>>, Error<ProviderError>> {
+        match &self.database {
+            Some(database) => database.load_logs().await.map_err(Error::Database),
+            None => Ok(vec![]),
+        }
+    }
+
+    async fn fetch_page(&self, filter: Filter) -> Result<Vec<Box<RawValue>>, Error<ProviderError>> {
+        let data: Result<Vec<Box<RawValue>>, ProviderError> = self
+            .provider
+            .provider()
+            .request("eth_getLogs", [&filter])
```

Comment:
Why do it manually instead of `provider.get_logs_paginated` ?

Reply:
Reverted it back to `provider.get_logs_paginated` but I'm not entirely sure which way is better.

Using `provider.get_logs_paginated` saves us some code (no more `Paginator` struct) and keeps things on higher abstraction layer (`Log` vs `Box<RawValue>`).

On the other hand, now we have to serialize `Log`s manually when storing them in db (see: https://github.com/worldcoin/signup-sequencer/pull/255/files#diff-abf9484cd24d8332d6292b4efa6f936372717653df66b551131ba70c9272d780R95). With us doing requests directly, we would store the original raw json log and deserialize once for stream consumers. Now, we get deserialized objects from `get_logs_paginated` and then serialize it again to store in db.